model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala

```diff
@@ -373,6 +371,56 @@ object ArrowInliner extends Logging {
     } yield values -> arrows
   }
 
+  //
+  private def collectionsToCanons(
+    tree: RawTag.Tree,
+    streamArgs: Map[String, VarModel]
+  ): RawTag.Tree = {
+    // collect arguments with stream type
+    // to exclude it from resolving and rename it with a higher-level stream that passed by argument
+    val streamsToRename = streamArgs.view.mapValues(_.name).toMap
+
+    if (streamsToRename.isEmpty) tree
+    else
+      tree
+        .map(_.mapValues(_.map {
+          // if an argument is a BoxType (Array or Option), but we pass a stream,
+          // change a type as stream to not miss `$` sign in air
+          case v @ VarRaw(name, baseType: CollectionType) if streamsToRename.contains(name) =>
+            v.copy(baseType = CanonStreamType(baseType.element))
+          case v => v
+        }))
+        .renameExports(streamsToRename)
+  }
+
+  // Canonicalize streams that passed to function as immutable collection variables
+  private def canonStreamVariables[S: Mangler](
+    args: ArgsCall
+  ): State[S, (List[CanonicalizeModel#Tree], Map[String, String])] = {
+    val streamToImmutableArgs = args.streamToImmutableArgs
+    for {
+      newStreamsCanon <- streamToImmutableArgs.values.toList.collect {
+        case vm @ VarModel(name, st: StreamType, _) =>
```

Comment:
Make `streamToImmutableArgs` return types along with model, so that you don't have to make a strange match here again to retrieve the type